sys/net/network_layer/ng_ndp/node/ng_ndp_node.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ */
+
+#include <stdint.h>
+
+#include "kernel_types.h"
+#include "net/ng_ipv6.h"
+#include "net/ng_ndp.h"
+#include "net/ng_pkt.h"
+#include "net/ng_pktqueue.h"
+
+#include "net/ng_ndp/internal.h"
+
+#include "net/ng_ndp/node.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+static char addr_str[NG_IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+static ng_pktqueue_t _pkt_nodes[NG_IPV6_NC_SIZE * 2];
+
+/**
+ * @brief   Allocates a node for the packet queue.
+ *
+ * @param[in] pkt   Packet to add to the packet queue.
+ *
+ * @return  A packet queue node.
+ */
+static ng_pktqueue_t *_alloc_pkt_node(ng_pktsnip_t *pkt)
+{
+    for (size_t i = 0; i < sizeof(_pkt_nodes) / sizeof(ng_pktqueue_t); i++) {
+        if ((_pkt_nodes[i].pkt == NULL) && (_pkt_nodes[i].next == NULL)) {
+            _pkt_nodes[i].pkt = pkt;
+
+            return &(_pkt_nodes[i]);
+        }
+    }
+
+    return NULL;
+}
+
+kernel_pid_t ng_ndp_node_next_hop_l2addr(uint8_t *l2addr, uint8_t *l2addr_len,
+                                         kernel_pid_t iface, ng_ipv6_addr_t *dst,
+                                         ng_pktsnip_t *pkt)
+{
+    ng_ipv6_addr_t *next_hop_ip = NULL, *prefix = NULL;
+
+#ifdef MODULE_NG_IPV6_EXT_RH
+    next_hop_ip = ng_ipv6_ext_rh_next_hop(hdr);
+#endif
+#ifdef MODULE_FIB
+    size_t next_hop_size;
+    uint32_t next_hop_flags = 0;
+    ng_ipv6_addr_t next_hop_actual; /* FIB copies address into this variable */
+
+    if ((next_hop_ip == NULL) &&
+        (fib_get_next_hop(&iface, next_hop_actual.u8, &next_hop_size,
+                          &next_hop_flags, (uint8_t *)dst,
+                          sizeof(ng_ipv6_addr_t), 0) >= 0) &&
+        (next_hop_size == sizeof(ng_ipv6_addr_t))) {
+        next_hop_ip = &next_hop_actual;
+    }
+
+#endif
+
+    if ((next_hop_ip == NULL)) {            /* no route to host */
+        if (iface == KERNEL_PID_UNDEF) {
+            /* ng_ipv6_netif_t doubles as prefix list */
+            iface = ng_ipv6_netif_find_by_prefix(&prefix, dst);
+        }
+        else {
+            /* ng_ipv6_netif_t doubles as prefix list */
+            prefix = ng_ipv6_netif_match_prefix(iface, dst);
+        }
+
+        if ((prefix != NULL) &&             /* prefix is on-link */
+            (ng_ipv6_netif_addr_get(prefix)->flags &
+             NG_IPV6_NETIF_ADDR_FLAGS_NDP_ON_LINK)) {
+            next_hop_ip = dst;
+#ifdef MODULE_FIB
+            /* We don't care if FIB is full, this is just for efficiency
+             * for later sends */
+            fib_add_entry(iface, (uint8_t *)dst, sizeof(ng_ipv6_addr_t), 0,
```

Comment:
(Moved function, but why is that a senseless entry? It shortens the search for the next hop, since the FIB is the first place the function looks for it.)